lib/target/generic_cpp/generic_cxx_codegen.cc

```diff
@@ -313,7 +313,111 @@ std::string GenericCXXCodeGen::EmitType(const halo::Type& type) {
 }
 
 static std::string DeclAsExtern(const std::string& str) {
-  return "extern \"C\" {   " + str + ";\n}\n";
+  return "extern \"C\" {\n" + str + ";\n};\n";
+}
+
+std::string GenericCXXCodeGen::GenerateTestFunc(const Function& func,
+                                                const std::string& func_decl,
+                                                const Instruction& ret_inst) {
+  std::ostringstream oss;
+  oss << "#include \"unittests.h\"\n\n";
+
+  if (func.IsEntryFunction()) {
+    if (opts_.dialect == Dialect::CXX_11) {
+      oss << DeclAsExtern(func_decl);
+    }
+    oss << "int main(int argc, char** argv) {\n";
+    oss << "  std::string test_case_dir = argv[argc - 1];\n";
+    oss << "  std::string device_name = argv[argc - 2];\n";
+    oss << "  int data_set_id = atoi(argv[argc - 3]);\n";
+    oss << "  double thre = strtod(argv[argc - 4], NULL);\n\n";
+    oss << "  std::vector<const void*> inputs;\n";
+    oss << "  std::vector<const void*> output_refs;\n";
+    oss << "  std::vector<void*> outputs;\n\n";
+    oss << "  UnitTests unittests;\n";
+    int32_t i = 0;
+    std::string data_type;
+    // load input data
+    for (auto& arg : func.Args()) {
+      auto& type = arg->GetResultType();
+      data_type.clear();
+      switch (type.GetDataType()) {
+        case DataType::FLOAT32:
+          data_type = "float";
+          break;
+        case DataType::INT32:
+          data_type = "int32_t";
+          break;
+        case DataType::UINT32:
+          data_type = "uint32_t";
+          break;
+        case DataType::INT64:
+          data_type = "int64_t";
+          break;
+        default:
+          HLCHECK(0);
+      }
+      oss << "  std::vector<" << data_type << "> in_" << i
+          << " = unittests.LoadInData<" << data_type << ">("
+          << "test_case_dir, data_set_id, " << i << ");\n";
+      oss << "  inputs.push_back(in_" << i << ".data());\n";
+      i++;
+    }
+
+    i = 0;
+    // load reference data and declare output
+    for (auto& out : ret_inst.GetOperands()) {
+      const auto& type = out.GetType();
+      const auto elem_nums = type.GetTotalNumOfElements();
+      data_type.clear();
+      switch (type.GetDataType()) {
```

Comment:
lines 344-359 and lines 373-388 are the same. Can you factor them out?

Reply:
OK， I have extracted  a lamda function "auto convert_data_type = [](DataType dtype)"